[AIRFLOW-558] Add Support for dag.backfill=(True|False) Option

btallman opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- \* https://issues.apache.org/jira/browse/AIRFLOW-558 *

Add Support for dag.backfill=(True|False) Option

Added a dag.backfill option and modified the scheduler to look at
the value when scheduling DagRuns (by moving dag.start_date up to
dag.previous_schedule), and added a config option scheduler.backfill_by_default (defaults to True)
that allows users to set this to False for all dags modifying the existing DAGs

In addition, we added a test to jobs.py (test_dag_backfill_option)

bolkedebruin commented on docs/scheduler.rst:
```diff
@@ -28,6 +30,27 @@ To start a scheduler, simply run the command:
+``DAG Runs`` the scheduler will begin at the first available interval after start date that has
+not already run, and schedule forward from there.
+
+*Added in Airflow 1.8*
```
Please dont mention this here. Documentation will be release together with Airflow itself, so there is no need.

bolkedebruin commented on airflow/models.py:
```diff
@@ -1020,11 +1020,45 @@ def are_dependents_done(self, session=None):
+
+    @property
+    @provide_session
+    def previous_dagrun(self, session=None):
```
I would expect this to be part of class DagRun not of TaskInstance.

btallman replied:
I put it here, because `ti.previous_dagrun()` seems cleaner than `dr.previous_dagrun(ti)`

It is a helper function, as we could also leverage something like:

```
tip = ti.previous_ti()
if tip:
    drp = tip.get_dagrun()
```

bolkedebruin replied:
Dr = Ti.get_dagrun()
Previous = dr.previous_dagrun()

DagRuns are the basis where to work from. TIs should only handle stuff within their own unit. Using previous here disregards that.

btallman replied:
OK, cleaned it up... Good idea. Interestingly, there was no dr.get_previous_dagrun, so I added that.

btallman replied:
A decision on style for ti.previous_ti vs ti.get_previous_ti() is needed as well, but that isn't something that I introduced.

ldct commented on tests/jobs.py:
```diff
@@ -851,3 +851,120 @@ def test_dag_with_system_exit(self):
+        """
+
+        now = datetime.datetime.now()
+        three_hours_ago_to_the_hour = (now - datetime.timedelta(hours=6)).replace(minute=0, second=0, microsecond=0)
```
this should be named `six_hours_ago_to_the_hour`

btallman replied:
@zodiac - Missed that one when we switched to testing a longer period. Good catch.

r39132 replied:
Agree with @zodiac

ldct commented on tests/jobs.py:
```diff
@@ -851,3 +851,120 @@ def test_dag_with_system_exit(self):
+        now = datetime.datetime.now()
+        three_hours_ago_to_the_hour = (now - datetime.timedelta(hours=6)).replace(minute=0, second=0, microsecond=0)
+        now_minus_three_minutes = now - datetime.timedelta(minutes=3)
+        now_minus_two_hours_and_three_minutes = now_minus_three_minutes - datetime.timedelta(hours=2)
```
These can be named `three_minutes_ago` and `two_hours_and_three_minutes_ago` for consistency

btallman replied:
@zodiac - That would be better, not sure if this will make it into a PR.

r39132 replied:
Agree with @zodiac

ldct commented on airflow/configuration.py:
```diff
@@ -458,6 +464,12 @@ def run_command(command):
+# Default behavior is unchanged
+# Command Line Backfills still work, but the scheduler
+# will not do backfills if this is false.
+backfill_by_default = true
```
This section doesn't need to be here, the default value (as in `DEFAULT_CONFIG`) of `backfill_by_default` will be used if this entry

Even if we want to include it for clarity the comments should be dropped

btallman replied:
@zodiac - Actually, test config is used by the Unit Tests, and so it does seem to be required. It would be a good idea to have the Test Config inherit from the Default config, and it looks like it does. However, it has not been my experience that this works. In fact the tests failed until I added it here.

ldct replied:
that's strange, I read the configparser code and thought test config selectively overrides default config... will need to investigate

btallman replied:
Yes, that is what it looks like to me as well, first default, then test. However, due to the day job, I didn't track that one down.

I had the same issue with the config.getboolean(), and decided to leave it for now and circle back once our showstoppers where fixed. In addition, I wanted to keep this PR clean and not involve extraneous fixes.

btallman replied:
Turns out true wasn't being read as True and false wasn't being read as False

ldct commented on airflow/jobs.py:
```diff
@@ -721,6 +721,25 @@ def create_dag_run(self, dag, session=None):
+            if not dag.backfill:
+                # The logic is that we move start_date up until
+                # one period before, so that datetime.now() is AFTER
+                # the period end, and the job can be created...
```
I guess this works, but it seems wrong to modify the `dag` argument (which was previously being used "read-only"). I think `create_dag_run` would be cleaner if we refactored it to avoid having to do this.

btallman replied:
The way the scheduler is written, I'm not sure there is a clean way to do this otherwise. @bolkedebruin and I talked this over before I started on this and we landed on this as the only clean way to do this. In the same way that dependencies have been refactored recently, scheduler needs a refactor. However, I think that is best left to 2.0 and this is a clean way to keep the scheduler from running backfills with the least code change (less chance of bugs).

ldct replied:
Hmm ok... even in that case I think it is still possible to avoid modifying `dag` by using `new_start` instead of `dag.start_date` in the rest of the dagrun scheduling computation

btallman replied:
@zodiac - I (and Apigee) chose this way, as it affects less code and is less prone to copy/paste errors.

ldct replied:
I know this changes less code, I just don't think the trade-off is worth it here.

@bolkedebruin any thoughts here?

bolkedebruin replied:
@btallman I think I agree with @zodiac, just because we store the start_date in the DB and it is kind of assumed then it is immutable. On the other hand it does add to the complexity.

What you could do is use a local start_date that is set to dag.start_date at the beginning of the function and use that down the line?

ldct commented on tests/jobs.py:
```diff
@@ -851,3 +851,120 @@ def test_dag_with_system_exit(self):
+
+        default_backfill = configuration.getboolean('scheduler', 'backfill_by_default')
+        # Test configs have backfill by default ON
+        if default_backfill == 'true':
```
these shouldn't be necessary because the `getboolean` call will only return booleans (https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.getboolean)

btallman replied:
@zodiac - After reviewing the results of the tests in further detail, it is possible that the false failures were due to basic limitations of the travis build process. I will verify and act accordingly.

ldct commented on tests/jobs.py:
```diff
@@ -851,3 +851,120 @@ def test_dag_with_system_exit(self):
+        self.assertEqual(dag1.backfill, True)
+
+        # Change the default to False and validate
+        configuration.set('scheduler', 'backfill_by_default', 'false')
```
we shouldn't modify configuration in tests...I don't see why it's necessary (since we pass `backfill=False` explicitly) and I don't know if other code depends on it not to be modified

btallman replied:
The whole point of this is to test the config, one way, then the other. The tests should reset the config each time, so it isn't an issue.

Not sure how we could test two config values without changing the config though. Open to ideas...

bolkedebruin commented on airflow/models.py:
```diff
@@ -1022,12 +1022,25 @@ def are_dependents_done(self, session=None):
+
+        else:
+
+            dr = self.get_dagrun()
```
This should pass a session, now we create two instead of one.

btallman replied:
Ahh, didn't realize I could do that, but obvious now...

bolkedebruin commented on airflow/models.py:
```diff
@@ -1022,12 +1022,25 @@ def are_dependents_done(self, session=None):
+                TaskInstance.dag_id == self.dag_id,
+                TaskInstance.task_id == self.task.task_id,
+                TaskInstance.execution_date ==
+                self.task.dag.previous_schedule(self.execution_date),
```
This should also be rewritten to use dag run as a base.

btallman replied:
This is the existing code base. Let me look at it's logic and see about adding to DR, although this PR is getting long in the tooth (old)...

bolkedebruin replied:
I know, but as you are changing the function anyway it would be nice to have it fixed properly.

bolkedebruin commented on airflow/models.py:
```diff
@@ -1022,12 +1022,25 @@ def are_dependents_done(self, session=None):
+                self.task.dag.previous_schedule(self.execution_date),
+            ).first()
+
+        else:
```
This is not required. You can leave "else" out here.

btallman replied:
I find the else to be more readable, but happy to drop it and the indent...

bolkedebruin replied:
Yes please!

mistercrunch commented on airflow/models.py:
```diff
@@ -2548,6 +2563,7 @@ def __init__(
             dagrun_timeout=None,
             sla_miss_callback=None,
             orientation=configuration.get('webserver', 'dag_orientation'),
+            backfill=configuration.getboolean('scheduler', 'backfill_by_default'),
```
`allow_scheduler_catchup=configuration.getboolean('scheduler', 'allow_scheduler_catchup_default')`

mistercrunch commented on airflow/models.py:
```diff
@@ -2532,6 +2545,8 @@ class DAG(BaseDag, LoggingMixin):
     :type sla_miss_callback: types.FunctionType
     :param orientation: Specify DAG orientation in graph view (LR, TB, RL, BT)
     :type orientation: string
+    :param backfill: Perform backfills (or only run latest)? Defaults to True
```
clearly describes how it affects the scheduler behavihor around DagRun creation

mistercrunch commented on airflow/configuration.py:
```diff
@@ -350,6 +350,12 @@ def run_command(command):
 
 child_process_log_directory = /tmp/airflow/scheduler/logs
 
+# Turn off scheduler backfills by setting this to False.
```
```
# Default setting for `DAG.allow_scheduler_catchup` [clearly describe how it affects DagRun creation]
allow_scheduler_catchup_default = True
```

bolkedebruin commented on airflow/models.py:
```diff
@@ -2730,10 +2739,6 @@ def concurrency(self, value):
         self._concurrency = value
 
     @property
-    def description(self):
```
This was added in a recent PR, please don't remove it

btallman replied:
Yeah, that was a rebase error!!! Sorry.

bolkedebruin commented on airflow/models.py:
```diff
@@ -2570,7 +2579,6 @@ def __init__(
         self._concurrency = concurrency
         self._pickle_id = None
 
-        self._description = description
```
Idem as below.

bolkedebruin commented on airflow/models.py:
```diff
@@ -3822,6 +3827,26 @@ def get_dag(self):
+
+        return session.query(DagRun).filter(
+            DagRun.dag_id == self.dag_id,
+            DagRun.execution_date < self.execution_date
```
While it is outside the scope of this PR, this is potentially a quite expensive operation on the DB.

bolkedebruin commented on airflow/models.py:
```diff
@@ -2534,11 +2541,12 @@ class DAG(BaseDag, LoggingMixin):
 
     def __init__(
             self, dag_id,
-            description='',
```
Please leave description in. See below.

bolkedebruin commented on airflow/models.py:
```diff
@@ -1022,12 +1022,21 @@ def are_dependents_done(self, session=None):
     @provide_session
     def previous_ti(self, session=None):
         """ The task instance for the task that ran before this task instance """
```
Much cleaner. Thanks!

btallman replied:
Going to have to dig into this. Apparently, moving to a DR based previous_ti causes issues when NOT using a DR (backfill). Finally figure that out this afternoon. Will attempt to fix tonight.

bolkedebruin commented on airflow/models.py:
```diff
@@ -3828,6 +3848,29 @@ def get_dag(self):
+        ).first()
+
+    @provide_session
+    def get_previous_scheduled_dagrun(self, session=None):
```
I'm not sure if I understand the difference between get_previous_dagrun and get_previous_scheduled_dagrun . Should get_previous_dagrun also not return the same as get_previous_scheduled_dagrun? Or do you want to distinguish between backfilled and intervalled dag runs?

Please rename to "get_previous_scheduled" it is already part of DagRun.

bolkedebruin commented on airflow/models.py:
```diff
@@ -1023,12 +1023,28 @@ def are_dependents_done(self, session=None):
+        dag = self.task.dag
+        if dag:
+            dr = self.get_dagrun(session=session)
+            if not dr:
```
I understand why you are doing this, although the comment does not reflect the requirement I think. Please mark it as FIXME (it should be removed in the future / throw an exception)

bolkedebruin commented on airflow/models.py:
```diff
@@ -1023,12 +1023,28 @@ def are_dependents_done(self, session=None):
+                if not previous_scheduled_date:
+                    return None
+                else:
+                    return TaskInstance(task=self.task, execution_date=previous_scheduled_date)
```
This means that whether or not there is a "previous_scheduled_date" (i.e. there won't be one by convention before start_date) one always gets a TaskInstance. I don't think that should be the case.

bolkedebruin commented on airflow/models.py:
```diff
@@ -3828,6 +3848,29 @@ def get_dag(self):
         return self.dag
 
     @provide_session
+    def get_previous_dagrun(self, session=None):
```
please rename to "get_previous", it is already part of DagRun (or even make it a property)

bolkedebruin commented on airflow/configuration.py:
```diff
@@ -485,6 +493,8 @@ def run_command(command):
 authenticate = true
 max_threads = 2
+catchup_by_default = True
+scheduler_zombie_task_threshold = 300
```
This is a duplicate from above

bolkedebruin reviewed (changes requested):
Some minor things. I will start integration testing now.

btallman committed:
Add Support for dag.catchup=(True|False) Option

Added a dag.catchup option and modified the
scheduler to look at the value when scheduling DagRuns
(by moving dag.start_date up to dag.previous_schedule),
and added a config option catchup_by_default
(defaults to True) that allows users to set this to False for all
dags modifying the existing DAGs

In addition, we added a test to jobs.py
(test_dag_catchup_option)
